Mech_lab5_1/src/main/java/agh/ics/oop/AbstractWorldMap.java

```diff
@@ -0,0 +1,58 @@
+package agh.ics.oop;
+
+import java.util.ArrayList;
+import java.util.List;
+
+abstract public class AbstractWorldMap implements IWorldMap {
+    private MapVisualizer map = new MapVisualizer(this);
+    protected List<Animal> animals = new ArrayList<>();
+    public abstract Vector2d getUpperRight();
+    public abstract Vector2d getLowerLeft();
+
+    public String toString() {
+        return this.map.draw(getLowerLeft(), getUpperRight());
+    }
+
+    @Override
+    public boolean isOccupied(Vector2d position) {
+        return (this.objectAt(position) != null);
+    }
+
+    public Object objectAt(Vector2d position) {
+
+        return animals.stream()
+                .filter(animal -> animal.isAt(position))
+                .findFirst().orElse(null);
+    }
+
+    @Override
+    public boolean canMoveTo(Vector2d position) {
+        if(this.isOccupied(position)) {
+            return false;
+        }
+        return position.precedes(getUpperRight()) && position.follows(getLowerLeft());
```

Comment:
to akurat jest zbędne i niepoprawne - w przypadku `GrassField` nie chcemy żeby fakt dojścia do krawędzi nas ograniczał, mapa rozszerza się dowolnie. Generalnie całe `canMoveTo()` zostawiłbym abstrakcyjne, każda z map inaczej to rozpatruje.